### ISSUE TITLE
Improve documentation of loss functions

### ISSUE THREAD

**Issue by tomaarsen** [UKPLab/sentence-transformers#2443]
Hello!

## Context
This issue describes documentation improvements that I would like to see.

## Details
Sentence Transformers, and the embedding space as a whole, features a lot of different loss functions, each designed with different purposes and with different inputs. Ideally, I would like to update the docstrings of all loss functions such that they contain more structured data. E.g. they have a table with the valid input formats, like for `MultipleNegativesRankingLoss`:

| Texts | Label |
|-|-|
| (anchor, positive) pairs | - |
| (anchor, positive, negative) triplets | - |

Or for `SoftmaxLoss`:
| Texts | Label |
|-|-|
| (sentence_A, sentence_B) pairs | class |

Some ideas:
- [ ] Input data format.
- [ ] Relations to other loss functions (e.g. "equivalent to `MultipleNegativesRankingLoss` but with caching" or "equivalent to `MSELoss` but with a margin")
- [ ] Purpose, if relevant.
- [ ] A code snippet.

With other words, currently the [loss docs](https://sbert.net/docs/package_reference/losses.html) can't be feasibly understood by someone who is experimenting with the project. I would like to change that.

cc @bwanglzu @ir2718 @johneckberg @aamir-s18 as I know you're interested in my TODO list. 

- Tom Aarsen

**Comment by ir2718** [comment]
@tomaarsen Could you elaborate a bit more on what you mean by purpose?

**Comment by tomaarsen** [comment]
Definitely!
Some loss functions were specifically introduced for certain purposes, e.g. for knowledge distillation or NLI. I don't strictly need "purpose" as that might indeed be too vague or restrictive (after all, some losses may have been introduced for NLI, but work wonders for IR as well).

So I'll explain the goal that I see with it instead: for users to be able to filter out loss functions that certainly won't be useful for them. For example, if you don't have hard negatives or similarity scores.

- Tom Aarsen
